trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+                    if first.is_data() && first.data()?.len() <= 1 {
+                        TransactionOutcome::StatusCode(first.as_val()?)
+                    } else {
+                        TransactionOutcome::StateRoot(first.as_val()?)
+                    }
+                },
+            }),
+            _ => Err(anyhow!("Invalid encoded rlp length.")),
+        }
+    }
+}
+
+impl LegacyReceipt {
+    pub fn new(
+        outcome: TransactionOutcome,
+        cumulative_gas_used: U256,
+        logs: Vec<LogEntry>,
+    ) -> Self {
+        LegacyReceipt {
+            cumulative_gas_used,
+            log_bloom: logs.iter().fold(Bloom::default(), |mut b, l| {
+                b.accrue_bloom(&l.bloom());
+                b
+            }),
+            logs,
+            outcome,
+        }
+    }
+    pub fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        match rlp.item_count()? {
+            3 => Ok(LegacyReceipt {
+                outcome: TransactionOutcome::Unknown,
+                cumulative_gas_used: rlp.val_at(0)?,
+                log_bloom: rlp.val_at(1)?,
+                logs: rlp.list_at(2)?,
+            }),
+            4 => Ok(LegacyReceipt {
+                cumulative_gas_used: rlp.val_at(1)?,
+                log_bloom: rlp.val_at(2)?,
+                logs: rlp.list_at(3)?,
+                outcome: {
+                    let first = rlp.at(0)?;
+                    if first.is_data() && first.data()?.len() <= 1 {
+                        TransactionOutcome::StatusCode(first.as_val()?)
+                    } else {
+                        TransactionOutcome::StateRoot(first.as_val()?)
+                    }
+                },
+            }),
+            _ => Err(DecoderError::RlpIncorrectListLen),
+        }
+    }
+
+    pub fn rlp_append(&self, s: &mut RlpStream) {
+        match self.outcome {
+            TransactionOutcome::Unknown => {
+                s.begin_list(3);
+            }
+            TransactionOutcome::StateRoot(ref root) => {
+                s.begin_list(4);
+                s.append(root);
+            }
+            TransactionOutcome::StatusCode(ref status_code) => {
+                s.begin_list(4);
+                s.append(status_code);
+            }
+        }
+        s.append(&self.cumulative_gas_used);
+        s.append(&self.log_bloom);
+        s.append_list(&self.logs);
+    }
+}
+
+#[derive(Eq, Hash, Debug, Copy, Clone, PartialEq)]
+#[repr(u8)]
+pub enum TypedTxId {
+    EIP1559Transaction = 0x02,
+    AccessList = 0x01,
+    Legacy = 0x00,
+}
+
+impl TypedTxId {
+    pub fn try_from_wire_byte(n: u8) -> anyhow::Result<Self> {
+        match n {
+            x if x == TypedTxId::EIP1559Transaction as u8 => Ok(Self::EIP1559Transaction),
+            x if x == TypedTxId::AccessList as u8 => Ok(Self::AccessList),
+            x if (x & 0x80) != 0x00 => Ok(Self::Legacy),
+            _ => Err(anyhow!("Invalid byte selector for tx type.")),
+        }
+    }
+}
```

Comment:
We can remove the `#[repr(u8)]` and custom function in favor of `TryFrom<u8>`.

Reply:
Cool, I've implemented `TryFrom<u8>` instead of `try_from_wire_byte()` - but also leaving `#[repr(u8)]` in for now. I messed around with implementing `Into<u8>` ... but it makes the api a bit messy in the place where we do explicit conversion from `TransactionId` to `u8`...
eg.
`TransactionId::AccessList as u8` > `Into::<u8>::into(TransactionId::AccessList)`